src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -91,117 +99,199 @@ export class KernelExecution implements IDisposable {
-            for (const cellToExecute of codeCellsToExecute) {
-                traceCellMessage(cellToExecute.cell, 'Before Execute individual cell');
-                const result = this.executeIndividualCell(kernel, cellToExecute, notebook);
-                result.finally(() => this.cellExecutions.delete(cellToExecute.cell)).catch(noop);
-                const executionResult = await result;
-                traceCellMessage(cellToExecute.cell, `After Execute individual cell ${executionResult}`);
-                // If a cell has failed or execution cancelled, the get out.
-                if (
-                    cancelTokenSource.token.isCancellationRequested ||
-                    executionResult === vscodeNotebookEnums.NotebookCellRunState.Error
-                ) {
-                    traceInfo(
-                        `Cancel all remaining cells ${cancelTokenSource.token.isCancellationRequested} || ${executionResult}`
-                    );
-                    await Promise.all(
-                        codeCellsToExecute.map((cell) => {
-                            // Cancel and remove from map pending cells.
-                            return this.cancelCell(cell.cell);
-                        })
-                    );
-                    break;
-                }
-            }
+            await this.pendingExecution.chainExecution(async () => this.executeQueuedCells(notebookPromise, document));
+        } catch (ex) {
+            // Possible one cell failed, we need to stop everything else.
+            traceError(`Failed to execute notebook ${document.uri.toString()}, hence cancelling all`, ex);
+            await this.cancelAllCells(notebookPromise, document);
+            throw ex;
         } finally {
-            traceInfo(`Cancel all remaining cells after finally`);
-            await Promise.all(
-                codeCellsToExecute.map((cell) => {
-                    // Cancel and remove from map pending cells.
-                    return this.cancelCell(cell.cell);
-                })
-            );
             this.documentExecutions.delete(document);
             traceInfo('Restore notebook state to idle');
             await editor.edit((edit) =>
                 edit.replaceMetadata({ ...document.metadata, runState: vscodeNotebookEnums.NotebookRunState.Idle })
             );
         }
     }
+    /**
+     * Interrupts the execution of cells.
+     * If we don't have a kernel (Jupyter Session) available, then just abort all of the cell executions.
+     */
+    public async interrupt(document: NotebookDocument, notebookPromise?: Promise<INotebook>): Promise<InterruptResult> {
+        // Possible we don't have a notebook.
+        const notebook = notebookPromise ? await notebookPromise.catch(() => undefined) : undefined;
+        traceInfo('Interrupt kernel execution');
+        // Interrupt all the cells, & wait for all to complete.
```

Comment:
THis is still a little confusing. Maybe the comment should be:

```
// Get the execution promises for all of the pending cells and cancel all non executing cells.
```

At least I think that's what it _should_ be doing? However it looks like it cancels cells that are running too? Whereas a cancel doesn't cancel? 

Okay now I'm totally confused again.

Reply:
Shouldn't on interrupt waitForAllQueuedCells have a value of false? It's not cancelling everyting, it's interrrupting it all. Although it should cancel pending cells?

Reply:
Maybe it would be better to split it into more parts?

Part 1 - Get pending executions
Part 2 - Interrupt session and wait for pending executions
Part 3 - Cancel all non pending executions

Reply:
Let me try something else, I don't want to differentiate make a differentiation between cancelling a cell that has started vs cell that hasn't.
`CellExecution.cancel` should be agnostic of the state of the cell, its just a command for `cancel what ever you are doing`.

Trying to split it means we cancel pending cells & cancel non-pending cells separately when I don't think we should.
Will look to clean up this code.

Reply:
But agree, the code is a little confusing, too much going on

Reply:
What if you don't cancel the pending cells in the first step. Just get their list. Then cancel after the session interrupt.

Reply:
Let me know if the new code is better (thanks, i think its better - but that doesn't mean it can't get better), the feedback is most certainly useful & helping me make it more readable..

Reply:
Will try to move all of the code that does the following:
* Tracking pending cells
* Get next pending cell
* Cancel all pending cells

Basically thats more like a `PendingCellsStack` class that:
* Has a method to push into the stack
* Cancel all items in the stack
* Wait for all to complete
* Get next item from the stack

Not sure if it will be overkill, but will make the code in `KernelExecution` simpler, let me see ...

Reply:
Not going with that, the code is small.

Reply:
Yes definitely better.